### DISCUSSION TITLE
What should we cover next?

### DISCUSSION THREAD

**Discussion by mgeisler** [google/comprehensive-rust#32] [Ideas]
Hi all, thanks so much for all the PRs you've sent me and the team in the last 24 hours!

I've discussed with people a bit on [Reddit](https://www.reddit.com/r/rust/comments/zrs1of/new_rust_course_by_android_comprehensive_rust/) and on [Hacker News](https://news.ycombinator.com/item?id=34091271#34092751) about what topics we should still cover in the course. The main ones that come to my mind are

* Async Rust. We only cover traditional [thread-based concurrency](https://google.github.io/comprehensive-rust/concurrency.html) in the course.
* Interior or shared mutability. We don't cover `Cell` and `RefCell` in the course.

Other things on your mind?

One thing that we've discussed in the team  (and it was actually brought up during a class) is that it would be helpful to have a section with an opinionated guide to "Real World Rust". Something that tells you how to solve common problems, guides you towards useful libraries, and otherwise helps new developers on their way.

**Comment by mgeisler** [comment]
We also have #34 where we get a new section on how to call Rust from Java. I'm thinking we could add an ad-hoc section at the end of the course for such extra topics — extra in the sense that they might not fit in directly in the four days we currently use for the course.

**Reply by mgeisler** [reply]
We should include a page on https://docs.rs/crate/googletest/latest.

**Comment by Luosuu** [comment]
Actually I think we need more explanation about `why we need` and `when to use`. Rust is built with years of experience in software engineering but not every learner knows where these smart designs of Rust come from...

**Reply by mgeisler** [reply]
That is a good point! So you think we should put more emphasis on explaining why Rust is a good language to use? And perhaps also explain when it is not a good choice?

I wrote the course from the perspective of someone who already knows they want to use Rust, but you're right that we should also cover the other audience to make sure everybody can get up to speed.

**Reply by rbutarascu** [reply]
I would very much be interested in finding out why Rust is a good language, when to use it and when not to use it. Thank you!

**Reply by mgeisler** [reply]
Great, I created #132 for this!

**Comment by mgeisler** [comment]
@brandonpollack23, as we discussed on chat: we should add a section with suggested libraries for solving common tasks.

**Reply by brandonpollack23** [reply]
Yeah, as we mentioned in chat some suggestions I had were:

* clap command like parser (macro and non macro versions)
* build.rs scripting
* I also removed this section, but we might want to mention dealing with runtime evaluated constants with crates like [OnceCell](https://crates.io/crates/once_cell) or [lazy_static](https://crates.io/crates/lazy_static).
* closed source cargo package registry
* Various important other packages in blessed.rs
* Async Rust (Tokio/async std) and how futures work under the hood (Pinning and code generation with async await, etc.  There are some good talks to point people to for this)

And some lower priority ones:

* Unbounded lifetimes in unsafe
* Undefined behaviour in unsafe
* Custom allocators/switching allocators
* nostd rust intro and links to book

After I'm done with my review of the book and adding what I can that is useful I'll start on those and have them reviewed as well :)



[1]: https://rust-lang.github.io/rfcs/0246-const-vs-static.html

**Reply by mgeisler** [reply]
> * Async Rust (Tokio/async std) and how futures work under the hood (Pinning and code generation with async await, etc.  There are some good talks to point people to for this)

This is now being worked on by @djmitche, @rbehjati, and @sakex in #181.

**Comment by Mandragorian** [comment]
I believe a collection of common rust patterns might be useful. I was thinking of snippets of code that might not be _that_ common to be included in stdlib, but also are quite common that are worth knowing. Some examples I was thinking:

- Using an `enum` with two variants instead of `bool` for better readability
- Preventing struct literals by including a private field
- Using generics for dependency injection

This will probably not be a homogeneous list since "common but not that common". But I think having a small list of cool tricks can be motivational, plus it might give incentive to students to explore other aspects of the language on their own.

**Reply by sakex** [reply]
Some top of mind ideas

- const methods and const generics
- More standard collections like BTreeMap
- WASM
- build.rs
- Structure of a large project
- features

**Comment by djmitche** [comment]
That all sounds like a lot to pack into 4 days!

I wonder if we shouldn't have a collection of "things to explore next" that aren't actually covered in the class, but with links to start?

If that comes at the end, then a class that gets done witih an hour or two remaining could select one to explore together.

**Reply by mgeisler** [reply]
> That all sounds like a lot to pack into 4 days!
> 
> I wonder if we shouldn't have a collection of "things to explore next" that aren't actually covered in the class, but with links to start?

Yes, that would be a good model.

The way I'm starting to think of things is that we'll have a Rust Fundamentals part which is ~3 days in duration. Follow this, we will offer 1-day courses on a vararity of topics such as the ones @Mandragorian and @sakex mention above.

We're already starting on this path with your changes to Day 4 and with @qwandor's [bare-metal Rust day](https://google.github.io/comprehensive-rust/welcome-bare-metal.html).

**Comment by rbehjati** [comment]
On Thu, Apr 13, 2023 at 10:31 AM Martin Geisler ***@***.***>
wrote:

> That all sounds like a lot to pack into 4 days!
>
> I wonder if we shouldn't have a collection of "things to explore next"
> that aren't actually covered in the class, but with links to start?
>
> Yes, that would be a good model.
>
> The way I'm starting to think of things is that we'll have a Rust
> Fundamentals part which is ~3 days in duration. Follow this, we will offer
> 1-day courses on a vararity of topics such as the ones @Mandragorian
> <https://github.com/Mandragorian> and @sakex <https://github.com/sakex>
> mention above.
>
I like that way of splitting the course. It also helps with splitting the
work between instructors.
If a session ends early, I think it is more effective to use the remaining
time to go deeper into the already discussed topics, instead of starting a
new one. The other topics could potentially be additional extension
courses.

**Reply by mgeisler** [reply]
Exactly, it's much more flexible that way. I hope we can have regular sessions with the basics and then offer the more specialized topics on a more irregular basis.

**Comment by sakex** [comment]
I'm bootstraping the class on Webassembly in https://github.com/google/comprehensive-rust/issues/1012 .

[Here's the branch so far](https://github.com/google/comprehensive-rust/tree/webassembly)
